frame/balances/src/lib.rs

```diff
@@ -1270,47 +1238,41 @@ impl<T: Trait<I>, I: Instance> ReservableCurrency<T::AccountId> for Module<T, I>
 	) -> (Self::NegativeImbalance, Self::Balance) {
 		if value.is_zero() { return (NegativeImbalance::zero(), Zero::zero()) }
 
-		let old_account = Self::account(who);
-		let mut account = old_account.clone();
-
-		// underflow should never happen, but it if does, there's nothing to be done here.
-		let actual = cmp::min(account.reserved, value);
-		account.reserved -= actual;
-
-		Self::set_account(who, &account, &old_account);
-
-		(NegativeImbalance::new(actual), value - actual)
+		Self::mutate_account(who, |account| {
+			// underflow should never happen, but it if does, there's nothing to be done here.
+			let actual = cmp::min(account.reserved, value);
+			account.reserved -= actual;
+			(NegativeImbalance::new(actual), value - actual)
+		})
 	}
 
-	/// Move the reserved balance of one account into the free balance of another.
+	/// Move the reserved balance of one account into the balance of another, according to `status`.
 	///
 	/// Is a no-op if the value to be moved is zero.
 	fn repatriate_reserved(
 		slashed: &T::AccountId,
 		beneficiary: &T::AccountId,
 		value: Self::Balance,
-	) -> result::Result<Self::Balance, DispatchError> {
+		status: Status,
+	) -> Result<Self::Balance, DispatchError> {
 		if value.is_zero() { return Ok (Zero::zero()) }
 
 		if slashed == beneficiary {
```

Comment:
This needs to take into account `status`

Cannot make a reasonable multiline github suggestion but:

`if status == Status::Free` then do the logic below

`if status == Status::Reserved` then do nothing